src/webob/acceptparse.py

```diff
@@ -1005,11 +1006,15 @@ def acceptable_offers(self, offers):
         ]
         # sort by offer_index, ascending
         acceptable_offers_n_quality_factors.sort(key=lambda tuple_: tuple_[2])
+        # (stable) sort by specificity, descending
+        acceptable_offers_n_quality_factors.sort(
```

Comment:
Wouldn't this sort override the sort above?

Reply:
Great question, as it is a stable sort, it won't impact the ordering above unless the specificity has changed. For instance:

```
>>> a = create_accept_header('application/json, text/html')
>>> a.acceptable_offers(a, ['text/html', 'application/json'])
[('text/html', 1.0), ('application/json', 1.0)]
```

The ordering of text/html is still first, as order_index took precedence. However, if the specificity differs then specificity wins:

```
>>> a = create_accept_header('application/json, text/*')
>>> a.acceptable_offers(a, ['text/html', 'application/json'])
[('application/json', 1.0), ('text/html', 1.0)]
```
For completeness, in the event that quality scores outrank specificity, then quality is the final tie breaker:

```
>>> a = create_accept_header('application/json;q=0.1, text/*;q=1')
>>> a.acceptable_offers(a, ['text/html', 'application/json'])
[('text/html', 1.0), ('application/json', 0.1)]
```

Does this help?

Reply:
I'm actually surprised to see we are re-sorting the list several times instead of defining a single sort tuple. I must have missed that in the original review. @whiteroses can you help me understand this?

Reply:
The first sort is ascending, the second sort is descending. Which is why it is split into two. However, I did introduce the third sort in this pull request. You are correct that the code change could be one tuple sort:

```
    # (stable) sort by qvalue and then specificity, descending
    acceptable_offers_n_quality_factors.sort(
        key=lambda tuple_: (tuple_[1], tuple_[3]), reverse=True
    )
```

I'll make the change.

Reply:
Sorting using a tuple is fundamentally different from doing the separate sorts. My original question for @whiteroses was why it is being done as independent sorts.

Reply:
Also to be clear you can do ascending and descending sorts with the same tuple by negating some of the entries in the tuple. The decision to use a tuple versus independent sorts should not be dependent on that perceived limitation.

Almost certainly the independent sorts is not correct and it should’ve been done using a tuple so that the subsequent sorts could not move items across regions in the result.

Reply:
The two sorts can be combined into one by using a tuple and negating the value that needs to be sorted in descending order, which may be more efficient and/or elegant. I don't think the current behaviour is incorrect though, given that .sort() is a stable sort &mdash; could you help me understand why it is incorrect @mmerickel?

Reply:
A stable sort is about the input order versus the output order with respect to a specific sort key. It does not provide any such guarantee where items will end up with respect to one another when you change the sort key other than that they will not change position if the new sort key says they're fine. If the new sort key says they should change position then they will, without regard for the previous sort key. Maybe since you are doing the sorts from least-important to most-important it works out but I still don't think it is guaranteed.

I think the big thing for me whether the current code is correct or not is that a single sort would be 1) more efficient and 2) more obvious the intentions.

Reply:
Ok, will submit PR in about 10 days' time! (if it's not urgent)

Reply:
@mmerickel Submitted [PR](https://github.com/Pylons/webob/pull/401) to combine the sorts as you suggested. It is much more readable &mdash; not sure what I was thinking. Thanks!